packages/api/src/webhooks/cannon.ts

```diff
@@ -502,18 +489,44 @@ export default class WebhookCannon {
     }
   }
 
-  async recordingToVodAsset(
-    mp4RecordingUrl: string,
-    userId: string,
-    sessionId: string
+  async handleRecordingReadyChecks(
+    sessionId: string,
+    mp4Url: string,
+    isRetry = false
   ) {
-    const id = uuid();
-    const playbackId = await generateUniquePlaybackId(sessionId);
-
-    const session = await this.db.stream.get(sessionId);
+    const session = await this.db.stream.get(sessionId, {
+      useReplica: false,
+    });
     if (!session) {
-      throw new Error("session not found");
+      throw new UnprocessableEntityError("Session not found");
+    }
+
+    const { lastSeen, sourceSegments, isActive } = session;
+    const activeThreshold = Date.now() - USER_SESSION_TIMEOUT;
+    if (!lastSeen || !sourceSegments) {
+      throw new UnprocessableEntityError("Session is unused");
     }
+    if (lastSeen > activeThreshold) {
+      if (isRetry) {
+        throw new UnprocessableEntityError("Session is still active");
+      }
+      // there was an update after the delayed event was sent, so sleep a few
+      // secs (up to USER_SESSION_TIMEOUT) and re-check if it actually stopped.
+      await sleep(5000 + (lastSeen - activeThreshold));
+      return this.handleRecordingReadyChecks(sessionId, mp4Url, true);
+    }
+
+    if (typeof isActive === "boolean" && !isActive) {
+      throw new UnprocessableEntityError("Session recording already handled");
+    }
+    await this.db.stream.update(sessionId, { isActive: false });
+
+    await this.recordingToVodAsset(session, mp4Url);
```

Comment:
Is there an argument to be made for doing these in the opposite order in case `recordingToVodAsset` fails for whatever reason?

Reply:
I'm having to do the DB update first in order to avoid generating duplicate assets for the same recording. An alternative would be creating a unique index on the `source.sessionId` field but that would seem like overkill TBH. We need to sort out our transaction game here.